File: backend-go/internal/modules/system/service/print_template_service.go
package service

import (
	"drip-admin/backend-go/internal/common"
	"github.com/gin-gonic/gin"
)

func (s *Server) PrintTemplates(c *gin.Context) {
	page, err := common.ParsePage(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	status, err := parseOptionalInt(c, "status")
	if err != nil {
		common.HandleError(c, err)
		return
	}
	query := s.db.Model(&SysPrintTemplate{}).Where("deleted = 0")
	query = likeIfPresent(query, "code", c.Query("code"))
	query = likeIfPresent(query, "name", c.Query("name"))
	query = eqIfPresent(query, "status", status)
	query = query.Order("updated_at desc, id desc")
	var rows []SysPrintTemplate
	total, err := s.pageResult(query, page, &rows)
	successOrError(c, common.PageResult[SysPrintTemplate]{List: rows, Total: total, Page: page.Page, PageSize: page.PageSize}, err)
}

func (s *Server) PrintTemplate(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var row SysPrintTemplate
	if err := s.db.Where("id = ? and deleted = 0", id.Int64()).First(&row).Error; err != nil {
		common.HandleError(c, common.NewBusinessError(404000, "print template not found"))
		return
	}
	common.Success(c, row)
}

func (s *Server) CreatePrintTemplate(c *gin.Context) {
	var request PrintTemplateSaveRequest
	if err := common.BindJSON(c, &request); err != nil {
		common.HandleError(c, err)
		return
	}
	id, err := s.savePrintTemplate(0, request, true)
	successOrError(c, id, err)
}

func (s *Server) CopyPrintTemplate(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var request PrintTemplateCopyRequest
	if err := common.BindJSON(c, &request); err != nil {
		common.HandleError(c, err)
		return
	}
	if err := s.ensurePrintCodeAvailable(request.Code, nil); err != nil {
		common.HandleError(c, err)
		return
	}
	var source SysPrintTemplate
	if err := s.db.Where("id = ? and deleted = 0", id.Int64()).First(&source).Error; err != nil {
		common.HandleError(c, common.NewBusinessError(404000, "print template not found"))
		return
	}
	row := SysPrintTemplate{ID: common.NewID(), Code: request.Code, Name: request.Name, PaperType: source.PaperType, TemplateJSON: source.TemplateJSON, Status: intOrDefault(request.Status, 1)}
	successOrError(c, row.ID, s.db.Create(&row).Error)
}

func (s *Server) UpdatePrintTemplate(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	var request PrintTemplateSaveRequest
	if err := common.BindJSON(c, &request); err != nil {
		common.HandleError(c, err)
		return
	}
	_, err = s.savePrintTemplate(id, request, false)
	successOrError(c, nil, err)
}

func (s *Server) DeletePrintTemplate(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	successOrError(c, nil, s.db.Delete(&SysPrintTemplate{}, id.Int64()).Error)
}

func (s *Server) PrintTemplateStatus(c *gin.Context) {
	id, err := idParam(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	status, err := statusRequest(c)
	if err != nil {
		common.HandleError(c, err)
		return
	}
	successOrError(c, nil, s.db.Model(&SysPrintTemplate{}).Where("id = ? and deleted = 0", id.Int64()).Update("status", status).Error)
}

func (s *Server) savePrintTemplate(id common.Int64String, request PrintTemplateSaveRequest, create bool) (common.Int64String, error) {
	if err := common.RequiredString(request.Code, "code"); err != nil {
		return 0, err
	}
	if err := common.RequiredString(request.Name, "name"); err != nil {
		return 0, err
	}
	if err := common.RequiredString(request.PaperType, "paperType"); err != nil {
		return 0, err
	}
	if err := common.RequiredString(request.TemplateJSON, "templateJson"); err != nil {
		return 0, err
	}
	if request.Status == nil {
		return 0, common.NewBusinessError(400000, "status is required")
	}
	if err := s.ensurePrintCodeAvailable(request.Code, &id); err != nil && !create {
		return 0, err
	}
	if create {
		if err := s.ensurePrintCodeAvailable(request.Code, nil); err != nil {
			return 0, err
		}
		row := SysPrintTemplate{ID: common.NewID(), Code: request.Code, Name: request.Name, PaperType: request.PaperType, TemplateJSON: request.TemplateJSON, Status: *request.Status}
		return row.ID, s.db.Create(&row).Error
	}
	return id, s.db.Model(&SysPrintTemplate{}).Where("id = ? and deleted = 0", id.Int64()).Updates(map[string]any{
		"code":          request.Code,
		"name":          request.Name,
		"paper_type":    request.PaperType,
		"template_json": request.TemplateJSON,
		"status":        *request.Status,
	}).Error
}

func (s *Server) ensurePrintCodeAvailable(code string, currentID *common.Int64String) error {
	query := s.db.Model(&SysPrintTemplate{}).Where("code = ? and deleted = 0", code)
	if currentID != nil && *currentID != 0 {
		query = query.Where("id <> ?", currentID.Int64())
	}
	var count int64
	if err := query.Count(&count).Error; err != nil {
		return err
	}
	if count > 0 {
		return common.NewBusinessError(400000, "print template code already exists")
	}
	return nil
}

File: backend-python/app/modules/system/service/print_template_service.py
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.common.errors import BusinessError, not_found
from app.common.id import new_id
from app.common.pagination import PageQuery, PageResult
from app.modules.system.dto.print_template_request import PrintTemplateCopyRequest, PrintTemplateSaveRequest
from app.modules.system.entity import SysPrintTemplate


class PrintTemplateService:
    def __init__(self, db: AsyncSession) -> None:
        self.db = db

    async def list_templates(
        self,
        page: PageQuery,
        code: str | None,
        name: str | None,
        status: int | None,
    ) -> PageResult[dict]:
        filters = [SysPrintTemplate.deleted == 0]
        if code and code.strip():
            filters.append(SysPrintTemplate.code.like(f"%{code.strip()}%"))
        if name and name.strip():
            filters.append(SysPrintTemplate.name.like(f"%{name.strip()}%"))
        if status is not None:
            filters.append(SysPrintTemplate.status == status)
        total = await self.db.scalar(
            select(func.count()).select_from(SysPrintTemplate).where(*filters)
        ) or 0
        rows = (
            await self.db.scalars(
                select(SysPrintTemplate)
                .where(*filters)
                .order_by(SysPrintTemplate.updated_at.desc(), SysPrintTemplate.id.desc())
                .offset((page.page - 1) * page.page_size)
                .limit(page.page_size)
            )
        ).all()
        return PageResult(
            list=[self._vo(row) for row in rows],
            total=str(total),
            page=page.page,
            pageSize=page.page_size,
        )

    async def detail(self, template_id: int) -> dict:
        return self._vo(await self._find(template_id))

    async def save(self, template_id: int | None, request: PrintTemplateSaveRequest) -> str | None:
        await self._assert_code_available(request.code.strip(), template_id)
        if template_id is None:
            row = SysPrintTemplate(
                id=new_id(),
                code=request.code.strip(),
                name=request.name.strip(),
                paper_type=request.paper_type.strip(),
                template_json=request.template_json,
                status=request.status,
                deleted=0,
            )
            self.db.add(row)
            await self.db.commit()
            return str(row.id)
        row = await self._find(template_id)
        row.code = request.code.strip()
        row.name = request.name.strip()
        row.paper_type = request.paper_type.strip()
        row.template_json = request.template_json
        row.status = request.status
        await self.db.commit()
        return None

    async def copy(self, template_id: int, request: PrintTemplateCopyRequest) -> str:
        source = await self._find(template_id)
        await self._assert_code_available(request.code.strip(), None)
        row = SysPrintTemplate(
            id=new_id(),
            code=request.code.strip(),
            name=request.name.strip(),
            paper_type=source.paper_type,
            template_json=source.template_json,
            status=request.status,
            deleted=0,
        )
        self.db.add(row)
        await self.db.commit()
        return str(row.id)

    async def delete(self, template_id: int) -> None:
        row = await self._find(template_id)
        row.deleted = 1
        await self.db.commit()

    async def update_status(self, template_id: int, status: int | None) -> None:
        row = await self._find(template_id)
        row.status = 1 if status is None else status
        await self.db.commit()

    async def _find(self, template_id: int) -> SysPrintTemplate:
        row = await self.db.scalar(
            select(SysPrintTemplate).where(
                SysPrintTemplate.id == template_id,
                SysPrintTemplate.deleted == 0,
            )
        )
        if row is None:
            raise not_found("print template not found")
        return row

    async def _assert_code_available(self, code: str, template_id: int | None) -> None:
        query = select(SysPrintTemplate.id).where(
            SysPrintTemplate.code == code,
            SysPrintTemplate.deleted == 0,
        )
        if template_id is not None:
            query = query.where(SysPrintTemplate.id != template_id)
        if await self.db.scalar(query) is not None:
            raise BusinessError(409000, "print template code already exists")

    @staticmethod
    def _vo(row: SysPrintTemplate) -> dict:
        return {
            "id": str(row.id),
            "code": row.code,
            "name": row.name,
            "paperType": row.paper_type,
            "templateJson": row.template_json,
            "status": row.status,
            "createdAt": row.created_at,
            "updatedAt": row.updated_at,
        }
